File: chat/common/scalajs/src/main/scala/chat/common/UI.scala
package chat
package common

import util._

import scala.scalajs.js.JSConverters._
import scala.scalajs.js.Dynamic
import scala.scalajs.js.Dynamic.global

class UI(
    name: String,
    nameChanged: String => Unit,
    chatRequested: User => Unit,
    chatSelected: Chat => Unit,
    chatClosed: Chat => Unit,
    messageSent: String => Unit) {
  private object ui {
    var username: Dynamic = _
    var chats: Dynamic = _
    var users: Dynamic = _
    var nousers: Dynamic = _
    var chatlog: Dynamic = _
    var message: Dynamic = _
    var send: Dynamic = _
  }

  private val $ = global.$

  def updateUsers(users: Seq[User]): Unit = $ { () =>
    ui.users.find("> li:not(:last)").remove()

    if (users.nonEmpty) {
      ui.nousers.hide()
      ui.users prepend (users map { case user @ User(_, name) =>
        $("""<li/>""") append (
          $("""<a href="#"/>""") click { event: Dynamic =>
            event.preventDefault()
            chatRequested(user)
          }
          text name)
      }).toJSArray
    }
    else
      ui.nousers.show()
  }

  def updateChats(chats: Seq[Chat]): Unit = $ { () =>
    ui.chats.find("> li:not(:last)").remove()

    ui.chats prepend (chats map { case chat @ Chat(_, name, unread, active) =>
      val button = $("""<button type="button" class="close">×</button>""")
      val badge = $("""<span class="badge"/>""")
      if (unread > 0)
        badge text unread
      val item = $("""<li/>""")
      if (active)
        item addClass "active"
      item append (
        $("""<a href="#"/>""") click { event: Dynamic =>
          event.preventDefault()
          chatSelected(chat)
        }
        text name append (" ", badge, " ", button click { event: Dynamic =>
          event.preventDefault()
          chatClosed(chat)
        }))
    }).toJSArray
  }

  def updateMessages(messages: Seq[Message]): Unit = $ { () =>
    ui.chatlog.empty()

    ui.chatlog append (messages map { case Message(content, own) =>
      $("""<li/>""") addClass (if (own) "own" else "foreign") text content
    }).toJSArray
  }

  def clearMessage = $ { () =>
    ui.message.`val`("")
  }

  $ { () =>
    ui.username = global $ "#username"
    ui.chats = global $ "#chats"
    ui.users = global $ "#users"
    ui.nousers = global $ "#nousers"
    ui.chatlog = global $ "#chatlog"
    ui.message = global $ "#message"
    ui.send = global $ "#send"

    val placeholder = name

    ui.username attr ("placeholder", placeholder)

    ui.username on ("input", { () =>
      nameChanged(ui.username.`val`().toString match {
        case username if username.trim == "" => placeholder
        case username => username
      })
    })

    ui.username trigger "input"

    ui.nousers on ("click", { event: Dynamic => event.preventDefault() })

    ui.send on ("click", { () => messageSent(ui.message.`val`().toString) })

    ui.message on ("keyup", { event: Dynamic =>
      if (event.keyCode == 13) {
        event.preventDefault()
        ui.send trigger "click"
      }
    })
  }
}

File: chat/common/js/src/main/resources/util/ui.js
$(function() {
  "use strict"

  window.UI = function(
      nameChanged, chatRequested, chatSelected, chatClosed, messageSent) {
    var self = this
    self.storeLog = true

    self.name = "Anonymous"

    self.updateUsers = function(users) {
      ui.users.find("> li:not(:last)").remove()

      if (users.length != 0) {
        ui.nousers.hide()
        ui.users.prepend(users.map(function(user) {
          return $("<li/>").append(
            $("<a href=\"#\"/>").click(function(event) {
              event.preventDefault()
              chatRequested(user)
            })
            .text(user.name))
        }))
      }
      else
        ui.nousers.show()
    }

    self.updateChats = function(chats) {
      ui.chats.find("> li:not(:last)").remove()

      ui.chats.prepend(chats.map(function(chat) {
        var button = $("<button type=\"button\" class=\"close\">×</button>")
        var badge = $("<span class=\"badge\"/>")
        if (chat.unread > 0)
          badge.text(chat.unread)
        var item = $("<li/>")
        if (chat.active)
          item.addClass("active")
        return item.append(
          $("<a href=\"#\"/>").click(function(event) {
            event.preventDefault()
            chatSelected(chat)
          })
          .text(chat.name).append(" ", badge, " ", button.click(function(event) {
            event.preventDefault()
            chatClosed(chat)
          })))
      }))
    }

    self.updateMessages = function(messages) {
      ui.chatlog.empty()

      ui.chatlog.append(messages.toArray().reverse().map(function(message) {
        return $("<li/>").addClass(message.own ? "own" : "foreign").text(message.content)
      }))

      var last = ui.chatlog.children().get(-1)
      if (last)
        last.scrollIntoView(false)
    }

    self.clearMessage = function() {
      ui.message.val("")
    }

    var ui = {
      username: $("#username"),
      chats: $("#chats"),
      users: $("#users"),
      nousers: $("#nousers"),
      chatlog: $("#chatlog"),
      message: $("#message"),
      send: $("#send")
    }

    var placeholder = self.name

    ui.username.attr("placeholder", placeholder)

    ui.username.on("input", function() {
      var username = ui.username.val()
      var newusername = username.trim() == "" ? placeholder : username

      if (self.name != newusername) {
        self.name = newusername
        nameChanged(self.name)
      }
    })

    ui.username.trigger("input")

    ui.nousers.on("click", function(event) { event.preventDefault() })

    ui.send.on("click", function() { messageSent(ui.message.val()) })

    ui.message.on("keyup", function(event) {
      if (event.keyCode == 13) {
        event.preventDefault()
        ui.send.trigger("click")
      }
    })
  }
})
